File: uploadCollection.py
from db import destinationDb as ddb
import json
from dateutil.parser import isoparse
from datetime import datetime

collection_name = "query_res_backup"

with open("query_results.json", "r") as f:
    data = json.load(f)

def restore(obj):
    if isinstance(obj, str):
        try:
            # Convert ISO string back to datetime object
            dt = isoparse(obj)
            if isinstance(dt, datetime):
                return dt  # Firestore accepts datetime directly
            return obj
        except:
            return obj  # Not a timestamp string
    elif isinstance(obj, dict):
        return {k: restore(v) for k, v in obj.items()}
    elif isinstance(obj, list):
        return [restore(v) for v in obj]
    return obj

for doc_id, doc_data in data.items():
    print(f"Uploading doc {doc_id}")
    doc_data = restore(doc_data)
    ddb.collection(collection_name).document(doc_id).set(doc_data)

print(f"Uploaded {len(data)} documents to collection '{collection_name}'")

File: uploadCollection.js
const { destinationDb } = require('./db');
const fs = require('fs');

const collectionName = 'query_res_backup';

function restore(obj) {
  if (typeof obj === 'string') {
    try {
      // Check if it's an ISO date string
      const date = new Date(obj);
      if (!isNaN(date.getTime()) && obj.includes('T') && (obj.includes('Z') || obj.includes('+'))) {
        return date;
      }
      return obj;
    } catch {
      return obj;
    }
  } else if (obj !== null && typeof obj === 'object') {
    if (Array.isArray(obj)) {
      return obj.map(item => restore(item));
    } else {
      const restored = {};
      for (const [key, value] of Object.entries(obj)) {
        restored[key] = restore(value);
      }
      return restored;
    }
  }
  return obj;
}

const data = JSON.parse(fs.readFileSync('query_results.json', 'utf8'));

const batch = destinationDb.batch();
let batchCount = 0;
const promises = [];

for (const [docId, docData] of Object.entries(data)) {
  console.log(`Uploading doc ${docId}`);
  const restoredData = restore(docData);
  const docRef = destinationDb.collection(collectionName).doc(docId);
  batch.set(docRef, restoredData);
  batchCount++;

  if (batchCount >= 500) {
    promises.push(batch.commit());
    batchCount = 0;
  }
}

if (batchCount > 0) {
  promises.push(batch.commit());
}

Promise.all(promises)
  .then(() => {
    console.log(`Uploaded ${Object.keys(data).length} documents to collection '${collectionName}'`);
  })
  .catch(error => {
    console.error('Error uploading documents:', error);
  });